Fixing Bug#6362479 - dmSlave traversal

kapilash opened the pull request:
GetDMSlaves results in an erroneous output if the slave of a dm-device is also a dm-device. In such cases, we must traverse to the slaves till we reach a physical disk.
Two tests are introduced to test this functionality to test positive and negative scenarios.
Also included is a minor fix in diskdepend.cpp, which results in empty name under some rare circumstances.

niroyb commented on source/code/scxsystemlib/disk/scxlvmutils.cpp:
```diff
@@ -229,56 +260,79 @@ namespace SCXSystemLib {
+                   // we bail out if the loop count reaches maxLoopCount.
+                   out.str(L"");
+		   out << L"Crossed " << maxLoopCount << " while navigating links.";
+                   SCX_LOGHYSTERICAL(log, out.str());
```
This should be logged as an error. The message should be what caused the error. Something like the comment above.

kapilash replied:
Thanks. Fixed this.

niroyb commented:
Signed off

jeffaco commented on source/code/scxsystemlib/disk/scxlvmutils.cpp:
```diff
@@ -229,56 +261,83 @@ namespace SCXSystemLib {
+                   // cycles instead of relying on more traditional approaches.
+                   // This way the vast majority of positive scenarios will only pay the price of one integer increment and one comparison.
+                   out.str(L"");
+                   out << L"Crossed " << maxLoopCount << " while navigating links.";
```
Did you add this "just in case", or did you hit a case where this actually occurred?

I'm wondering if this should be a "log once" type thing (suspend multiple identical error messages)? And the error message leaves something to be desired. Sure, I know the threshold was exceeded, but how about something like:

L"Exceeded " << maxLoopCount << L" links while evaluating device " << dmDevice

That's better information to go on. But I still think you should do a "log once" type thing for a particular dmDevice to avoid needlessly filling the error log.

kapilash replied:
Yes. This is "just in case". I have not seen it happen anywhere. I was just not sure of the impact of an infinite loop on the server.
 I've fixed the log statement and made it "log once".

jeffaco commented on source/code/scxsystemlib/disk/diskdepend.cpp:
```diff
@@ -1438,6 +1438,7 @@ namespace SCXSystemLib
 
         if (SolarisPartitionNamePatternRegExPtr->IsMatch(path.GetFilename()))
         {    // remove the "p#", "s#", "p##" or "s##" from the end of the name
+
```
Any reason for this? In general, we don't do blank lines after a { ... we just go on the next line (indented).

jeffaco commented on test/code/scxsystemlib/disk/scxlvmutils_test.cpp:
```diff
@@ -1036,6 +1039,143 @@ class SCXLVMUtilsTest : public CPPUNIT_NS::TestFixture
+            SCXCoreLib::SCXFilePath blockFile = pathStr;
+
+            // to avoid getting unused variable warning
+            (void) options;
```
Prefer 1071-1072 to be at line 1067 (top of routine) to be clearer that the parameter is not used.

jeffaco commented on test/code/scxsystemlib/disk/scxlvmutils_test.cpp:
```diff
@@ -1036,6 +1039,143 @@ class SCXLVMUtilsTest : public CPPUNIT_NS::TestFixture
+
+            lines.clear();
+            lines.push_back(L"253:2");
+
```
Delete needless blank line before the }

jeffaco commented on test/code/scxsystemlib/disk/scxlvmutils_test.cpp:
```diff
@@ -1036,6 +1039,143 @@ class SCXLVMUtilsTest : public CPPUNIT_NS::TestFixture
+
+    void GetDMSlaves_SlavesWithDMEntries_TraversesToDevice()
+    {
+
```
Delete needless blank line after {. Blank lines are inserted for readability, in this case it's not all that useful.

jeffaco commented:
Most of this was nitpicking. This looks good to me, just consider how you want to log the error and if you want repetitive errors or just log once/device to not fill the log.

Signed off.

kapilash commented:
Thanks @jeffaco . I've removed the empty lines (sorry about those). I've fixed the log statement to indicate that the error is for which device. And also made it a "log once".
I've rebased against master and tested. Merging it.
